src/main/java/net/logstash/logback/util/ReusableJsonFormatterPool.java

```diff
@@ -0,0 +1,223 @@
+     *   are not returned to Jackson's internal memory pools.
+     *   This behavior is desired and makes the associated memory immediately reclaimable - which is what
+     *   we need since we are "running low in memory".
+     */
+    private volatile SoftReference<Deque<ReusableJsonFormatter>> formatters = new SoftReference<>(null);
+    
+
+    public ReusableJsonFormatterPool(CompositeJsonFormatter<Event> formatterFactory, int minBufferSize) {
+        this.formatterFactory = Objects.requireNonNull(formatterFactory);
+        this.minBufferSize = minBufferSize;
+    }
+
+    /**
+     * A reusable JsonFormatter holding a JsonFormatter writing inside a dedicated {@link ReusableByteBuffer}.
+     * Closing the instance returns it to the pool and makes it available for subsequent usage, unless the
+     * underlying {@link CompositeJsonFormatter.JsonFormatter} threw an exception during its use.
+     * 
+     * <p>Note: usage is not thread-safe.
+     */
+    public class ReusableJsonFormatter implements Closeable {
+        private final ReusableByteBuffer buffer;
+        private final CompositeJsonFormatter<Event>.JsonFormatter formatter;
+        private boolean recyclable = true;
+        
+        ReusableJsonFormatter(ReusableByteBuffer buffer, CompositeJsonFormatter<Event>.JsonFormatter formatter) {
+            this.buffer = buffer;
+            this.formatter = formatter;
+        }
+
+        /**
+         * Return the underlying buffer into which the JsonFormatter is writing.
+         * 
+         * @return the underlying byte buffer
+         */
+        public ReusableByteBuffer getBuffer() {
+            return buffer;
+        }
+        
+        /**
+         * Write the Event in JSON format into the enclosed buffer using the enclosed JsonFormatter.
+         * 
+         * @param event the event to write
+         * @throws IOException thrown when the JsonFormatter has problem to convert the Event into JSON format
+         */
+        public void write(Event event) throws IOException {
+            try {
+                this.formatter.writeEvent(event);
+                
+            } catch (IOException e) {
+                // Do not recycle the instance after an exception is thrown: the underlying
+                // JsonGenerator may not be in a safe state.
+                this.recyclable = false;
+                throw e;
+            }
+        }
+        
+        /**
+         * Close the JsonFormatter, release associated resources and return it to the pool.
+         */
+        @Override
+        public void close() throws IOException {
+            release(this);
+        }
+        
+        /**
+         * Dispose associated resources
+         */
+        protected void dispose() {
+            try {
+                this.formatter.close();
+                this.buffer.reset();
+                
+            } catch (IOException e) {
+                // ignore and proceed
+            }
```

Comment:
if formatter.close() throws, then buffer.reset() won't be called.  Probably should have two separate try/catch blocks, one for each, to ensure buffer.reset() is called even when formatter.close() throws.

Reply:
You are right... However, `buffer.reset()` does nothing more than clearing the buffer content and dropping any extra buffers that may have been allocated. Since the ReusableJsonFormatter is "dropped" (no reference to it anymore) not calling reset() doesn't have any consequences.

Closing the JsonFormatter however makes it unusable -> the underlying JsonGenerator is closed and will throw an exception if used again.

Your remark makes me think that the implementation would be clearer if:
- dispose() close the JsonFormatter and nullify the references to the enclosed JsonFormatter and ReusableByteBuffer
- add a guard to the other method throwing an IllegalStateException if the ReusableJsonFormatter is re-used after being disposed

Reply:
>  Since the ReusableJsonFormatter is "dropped" (no reference to it anymore) not calling reset() doesn't have any consequences.

Ok.  The current code "looks" like a bug.   I've seen other cases like this that lead to sneaky hard-to-debug leaks.  So, seeing the cleanup statements lumped together in a single try block jumps out at me as a red flag.

If it is ok for reset() to not be called, then remove the call to reset() OR add a comment above it that says it's ok to not call reset() if close() throws an exception.

If reset() _should_ be called, then split the try blocks.

Doing any of the above will make it clear that the code is "correct", and not a potential problem.  

> Closing the JsonFormatter however makes it unusable -> the underlying JsonGenerator is closed and will throw an exception if used again.

Ok, that is enough to satisfy my concern.  Thanks for clarifying.   Adding more defensive code will make the error message more clear if the formatter was reused after dispose, but I don't think that is necessary since this class is meant to be internal to the library.